src/coreclr/jit/gentree.cpp

```diff
@@ -12685,63 +12685,132 @@ GenTree* Compiler::gtFoldExpr(GenTree* tree)
 //
 // Notes:
 //    Checks for calls to Type.op_Equality, Type.op_Inequality, and
-//    Enum.HasFlag, and if the call is to one of these,
-//    attempts to optimize.
+//    Enum.HasFlag, as well as the helpers for casts,
+//    and if the call is to one of these, attempts to optimize.
 
 GenTree* Compiler::gtFoldExprCall(GenTreeCall* call)
 {
-    // Can only fold calls to special intrinsics.
-    if ((call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) == 0)
-    {
-        return call;
-    }
-
     // Defer folding if not optimizing.
     if (opts.OptimizationDisabled())
     {
         return call;
     }
 
-    // Fetch id of the intrinsic.
-    const CorInfoIntrinsics methodID = info.compCompHnd->getIntrinsicID(call->gtCallMethHnd);
-
-    switch (methodID)
+    // Is this an intrinsic we can fold?
+    if (call->IsSpecialIntrinsic())
     {
-        case CORINFO_INTRINSIC_TypeEQ:
-        case CORINFO_INTRINSIC_TypeNEQ:
+        // Fetch id of the intrinsic.
+        const CorInfoIntrinsics methodID = info.compCompHnd->getIntrinsicID(call->gtCallMethHnd);
+        switch (methodID)
         {
-            noway_assert(call->TypeGet() == TYP_INT);
-            GenTree* op1 = call->gtCallArgs->GetNode();
-            GenTree* op2 = call->gtCallArgs->GetNext()->GetNode();
+            case CORINFO_INTRINSIC_TypeEQ:
+            case CORINFO_INTRINSIC_TypeNEQ:
+            {
+                noway_assert(call->TypeGet() == TYP_INT);
+                GenTree* op1 = call->gtCallArgs->GetNode();
+                GenTree* op2 = call->gtCallArgs->GetNext()->GetNode();
+
+                // If either operand is known to be a RuntimeType, this can be folded
+                GenTree* result = gtFoldTypeEqualityCall(methodID, op1, op2);
+                if (result != nullptr)
+                {
+                    return result;
+                }
+                break;
+            }
+
+            default:
+                break;
+        }
+
+        // Check for a new-style jit intrinsic.
+        const NamedIntrinsic ni = lookupNamedIntrinsic(call->gtCallMethHnd);
+
+        if (ni == NI_System_Enum_HasFlag)
+        {
+            GenTree* thisOp = call->gtCallThisArg->GetNode();
+            GenTree* flagOp = call->gtCallArgs->GetNode();
+            GenTree* result = gtOptimizeEnumHasFlag(thisOp, flagOp);
```

Comment:
This is the new code optimization that you are adding. Right?

Reply:
Yes. Without it, the (new) cases in tests were not getting folded. That said, I did not find any diffs for this in the framework, so I would be 100% fine with backing this out.

Reply:
Note as well: there are some `DIV` and `MOD` helpers that could be folded here too, so that cases like [these](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRIpgDMFbGXZkA3kTJ6yAbQCyLBAAtOAEwCSvfgAoT5q7Z78A8jyacYKKgDlOaxh+WFgAcwBKAF1dfRUKXAA2MmBOTn4NCH4HSLIAXgA+MgQhOhYFQn0yOL0E2AQyYIQAEShxe0ja7W7qhrIIArJcEjlq6t79dmzcyvH9fuAhkbn5rJzO1fnJuoB2AbIAelStsgBfbu6E/h9wsgAZW7aOrqr9HTf5m5g7wfzh0ZrGqfcbTDaRU7Vb53Jb/FY7dazBEIzD7QbHYCnC6EM5AA=) would be optimized on 32 bit. But long divison / modulus are also very rare (may become not as rare if `BigInteger` is moved to use `long` internally).